src/edu/washington/escience/myria/operator/CSVFileScanFragment.java

```diff
@@ -200,99 +206,76 @@ protected TupleBatch fetchNextReady() throws IOException, DbException {
       }
 
       if (record != null) {
+
         /*
-         * Here, we read the first byte of the stream if this is not the first worker. This is simply an optimization.
-         * If the previous character is not a new line, we discard the line.
+         * Here, if we are on the last row, we make sure to read the entire row until we either hit a new line or until
+         * we have read the entire file (this is for the case where a single worker might be reading a single large row
+         * that was split among other workers). If we're at the last row and the last worker is reading, we just mark
+         * this final line as finished.
          */
-        if (lineNumber - 1 == 0 && partitionStartByteRange != 0) {
-          char currentChar =
-              (char)
-                  source.getInputStream(adjustedStartByteRange - 1, adjustedStartByteRange).read();
-          if (currentChar != '\n' && currentChar != '\r') {
-            discardedRecord = true;
+        if (onLastRow && !finishedReadingLastRow && !isLastWorker) {
+          long trailingStartByte = partitionEndByteRange + 1;
+          long trailingEndByte = trailingStartByte + byteOverlap;
+          boolean finalLineFound = false;
+
+          while (!finalLineFound) {
+            boolean withinByteRange = trailingEndByte < maxByteRange;
```

Comment:
Not sure that an extra variable is justified by a single use in the very next line.